client/protocol.py

```diff
@@ -0,0 +1,200 @@
+"""
+Protocol Client
+"""
+
+from __future__ import annotations
+import asyncio
+import logging
+
+import grpc
+from google.protobuf.internal.containers import RepeatedCompositeFieldContainer
+
+from api.curp.message_pb2_grpc import ProtocolStub
+from api.curp.message_pb2 import FetchClusterRequest, FetchClusterResponse, Member
+from api.curp.curp_command_pb2 import ProposeRequest, ProposeResponse, WaitSyncedRequest, WaitSyncedResponse
+from api.xline.xline_command_pb2 import Command, CommandResponse, SyncResponse
+from client.error import ResDecodeError, ProposeError, CommandSyncError, WaitSyncError, ExecuteError
+from api.curp.curp_error_pb2 import (
+    ProposeError as _ProposeError,
+    CommandSyncError as _CommandSyncError,
+    WaitSyncError as _WaitSyncError,
+)
+from api.xline.xline_error_pb2 import ExecuteError as _ExecuteError
+
+
+class ProtocolClient:
+    """
+    Protocol client
+
+    Attributes:
+        local_server_id: Local server id. Only use in an inner client.
+        state: state of a client
+        inner: inner protocol clients
+        connects: all servers's `Connect`
+    """
+
+    local_server_id: int
+    inner: list[ProtocolStub]
+    connects: RepeatedCompositeFieldContainer[Member]
+
+    def __init__(
+        self,
+        leader_id: int,
+        stubs: list[ProtocolStub],
+        connects: RepeatedCompositeFieldContainer[Member],
+    ) -> None:
+        self.local_server_id = leader_id
+        self.inner = stubs
+        self.connects = connects
+
+    @classmethod
+    def build_from_addrs(cls, addrs: list[str]) -> ProtocolClient:
+        """Build client from addresses, this method will fetch all members from servers"""
+
+        stubs: list[ProtocolStub] = []
+
+        for addr in addrs:
+            channel = grpc.insecure_channel(addr)
```

Comment:
We should close this channel after this function returns. Please use a context manager to make sure this channel will be closed.

Reply:
Python `grpc.Channel` object seem to internally manage the lifecycle of the resource, including connection establishment and closure. Don't need to close the channel manually

Reply:
FYI: https://grpc.io/docs/languages/python/quickstart/#update-the-client

Reply:
We shouldn't close this channel after this function returns. Because we need to maintain a channel to the node at all times to send requests, in the member conf change we will close the channel for the removed node.